cli/launchpad/model.go

```diff
@@ -0,0 +1,248 @@
+// Package launchpad file model.go contains all supported CustomResourceDefinition (CRD).
+//
+// Every CRD should have a `{kind}YAML` to denote the full CRD representation,
+// and a `{kind}SpecYAML` to denote the spec map inside the CRD
+//
+// Each `{kind}SpecYAML` should also implement stackable interface (synonymous to yaml.Unmarshaler)
+// to allow a stack like evaluation.
+package launchpad
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"regexp"
+)
+
+const (
+	folderNameMin = 3
+	folderNameMax = 30
+)
+
+var (
+	errValidationFailed = errors.New("validation failed")
+	tfNameRegex         = regexp.MustCompile("^[a-zA-Z][a-zA-Z\\d\\-\\_]*$")
+)
+
+// configYAML represents a fully qualified CRD that can be of any supported Kind.
+//
+// All CRDs are expected to have the following properties.
+type configYAML struct {
+	APIVersion string      `yaml:"apiVersion"`
+	Kind       crdKind     `yaml:"kind"`
+	Spec       interface{} `yaml:"spec"` // Placeholder, yielding for further evaluation
+	//Metadata   Metadata               `yaml:"metadata"`
+	Undefined map[string]interface{} `yaml:",inline"` // Catch-all for untended behavior
+
+	// Original YAML for backup purpose
+	rawYAML string
+}
+
+// commonConfigYAML alias configYAML for all CRDs to implement common properties.
+type commonConfigYAML configYAML
+
+// UnmarshalYAML evaluates common CRD attributes and dynamically parse the CRDs based on Kind field value.
+func (c *configYAML) UnmarshalYAML(unmarshal func(interface{}) error) (err error) {
+	type raw configYAML
+	if err := unmarshal((*raw)(c)); err != nil {
+		return err
+	}
+	switch c.Kind {
+	case KindCloudFoundation:
+		cf := cloudFoundationYAML{}
+		cf.APIVersion = c.APIVersion
+		cf.Kind = c.Kind
+		err = unmarshal(&cf)
```

Comment:
I'm very confused by this. I think you're mixing up the marshal / unmarshal terms.
https://en.wikipedia.org/wiki/Marshalling_(computer_science)

If this is really doing an unmarshal, it should be taking the serialized form of something and turning it into the inmemory representation, which in this case is a `cloudFoundationYAML` struct.

To be clear:
"Marshal" - Turn an in-memory struct into YAML
"Unmarshal" - Turn YAML into an in-memory struct.

I guess it's possible that the `unmarshal` func passed to UnmarshalYAML has some data reference in the closure but I can't find any uses of it to see.

This can also return the err directly.